File: solutions/111.minimum-depth-of-binary-tree/minimum-depth-of-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
// const minDepth = root => {
//   let counter = 0

//   const iter = (node, depth) => {
//     if (!node) return

//     if (counter && depth >= counter) return

//     if (!node.left && !node.right) {
//       counter = depth
//       return
//     }

//     if (node.left) {
//       iter(node.left, depth + 1)
//     }
//     if (node.right) {
//       iter(node.right, depth + 1)
//     }

//   }

//   iter(root, 1)
//   return counter
// }

const minDepth = root => {
  if (!root) return 0
  let min = Infinity
  let count = 1

  const walk = (node, count) => {
    if (!node.left && !node.right) {
      min = Math.min(min, count)
    }

    count++
    if (node.left) {
      walk(node.left, count)
    }

    if (node.right) {
      walk(node.right, count)
    }

  }

  walk(root, count)
  return min
}

File: solutions/111.minimum-depth-of-binary-tree/minimum-depth-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def minDepth(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root:
            return 0

        left = self.minDepth(root.left)
        right = self.minDepth(root.right)

        if left and right:
            return min(left, right) + 1

        return (left or right) + 1
